test/CreateTest.php

```diff
@@ -163,5 +163,14 @@ public static function provideGoodCredentials()
             ['guest', null],
         ];
     }
+
+	public function test_10_zero_retry_exception() {
```

Comment:
Should there also be a test that actually checks that no retry calls are made if retry_count = 0?

Reply:
Ideally we should be able to determine the following situations:

1. The connector performs an attempt to connect, which fails.
2. The connector performs an attempt to connect, which succeeds.

And trigger the following situations:

3. Make conditions that will lead to a connection fail.

`1` is doable. We can count successful connection attempts from tarantool side using an [on_connect trigger](https://www.tarantool.io/en/doc/1.10/reference/reference_lua/box_session/#box-session-on-connect) (with filtering by `box.session.type() == 'binary'`).

`2` requires to mock `php_stream_xport_create()` using LD_PRELOAD, implement simple tarantool server mock (binary protocol) or expose information about (un)successful attempts to connect from a connector instance.

`3` requires to be able to start / stop tarantool server from a test, mock `php_stream_xport_create()` using LD_PRELOAD, implement simple tarantool mock or provide an API to inject an artificial error into the code.

To be honest, I would prefer to avoid points from `2` and `3`, because they look as overkill for me. At least within the scope of this bugfix.

If we'll treat `retry_count` as attempts count after first unsuccessful one, then zero value means: try one time and don't try second time if there was an error. We should imitate an unsuccessful connection attempt.

If we'll treat `retry_count` as overall attempts count (like in current master), then zero value means: don't ever do any attempt to connect. This is simple to verify using an on_connect trigger.

Reply:
> To be honest, I would prefer to avoid points from 2 and 3, because they look as overkill for me. At least within the scope of this bugfix.

I agree that implementing such tooling is out of the scope of this ticket and I can't think of an easy way to test it, so feel free to ignore my initial comment.

Reply:
BTW, server-side iproto can be implemented with minimal effort on tarantool itself and this way seems to be quite flexible to use it in various testing scenarious. See example [here](https://lists.tarantool.org/pipermail/tarantool-patches/2020-May/016986.html), under `----` marker.

Maybe I'll use this way here or in some other project, that's interesting.